### ISSUE TITLE
Add the ability to specify `$node` in log formatter format string

### ISSUE THREAD

**Issue by lukebakken** [rabbitmq/rabbitmq-server#10921]
### Is your feature request related to a problem? Please describe.

In VESC-1100, one challenge we are facing is that the customer aggregates log lines from all nodes in their cluster into the log aggregation system, with no way to separate log lines by node. Thus, when providing logs, lines are interleaved between nodes.

Since this seems to be a common scenario in k8s environments, we should allow the ability to specify `$node` in the format statement for log lines.

### Describe the solution you'd like

Allow a format string like this:

```
log.console.formatter.plaintext.format = $node $time [$level] $pid $msg
```

### Describe alternatives you've considered

You _can_ already specify a constant string in the format string:

```
log.console.formatter.plaintext.format = rmq0 $time [$level] $pid $msg
```

However, this means that the configuration file must be templated before the node starts.

### Additional context

See VESC-1100

**Comment by Vishal-770** [comment]
Hi @lukebakken,

I traced through `rabbitmq_prelaunch` and the logging lifecycle to look into supporting `$node` in the plaintext log formatters.

Right now, `rabbit_prelaunch_early_logging:prepare_fmt_format/1` already extracts `$node` into the atom `node` via the existing regex. The reason it prints `<unknown node>` today is that `rabbit_logger_text_fmt:format_var/3` only matches `level` and `time`, falling back to `maps:get(Var, Meta, undefined)` for anything else. Because `node` is not part of event metadata, it falls through to `<unknown node>`.

We can handle this cleanly by adding a dedicated clause in `rabbit_logger_text_fmt:format_var/3`:

```erlang
format_var(node, #{meta := Meta}, _Config) ->
    Node = maps:get(node, Meta, node()),
    NodeStr = rabbit_data_coercion:to_list(Node),
    io_lib:format("~ts", [NodeStr]);
```

A few details on why it is shaped this way and the edge cases it handles:

* **Kubernetes hostnames and quoting**: In Kubernetes, pod hostnames often look like `rabbit@rabbitmq-0.rabbitmq-nodes.default.svc.cluster.local`. Because `node()` returns an atom, using Erlang's default `~tp` atom formatting would wrap the node name in single quotes (`'rabbit@rabbitmq-0...'`) due to the hyphens and dots. Coercing it with `rabbit_data_coercion:to_list/1` and formatting with `~ts` ensures the output remains clean and unquoted
* **Early boot safety**: Prior to Erlang distribution starting up in early prelaunch, `node()` returns the atom `nonode@nohost`. This ensures log statements emitted during prelaunch will format safely without crashing
* **Subsystem metadata precedence**: Checking `maps:get(node, Meta, node())` ensures that if an internal subsystem (such as Ra or a custom plugin) attaches explicit node metadata to a log event, that value takes precedence over the local node name
* **Conditional formatting**: In `format_prefix/4`, conditional blocks like `${node?[$node] :}` currently check `maps:get(Var, Meta, undefined)`. We should update that check so `Var =:= node` recognizes the presence of `node()` even when it is absent from `Meta`
* **Scope across handlers**: Because `rabbit_logger_text_fmt` is shared across console, file, syslog, and exchange loggers, this single change enables `$node` across all four plaintext formatters (`log.console`, `log.file`, `log.syslog`, and `log.exchange`)

I also considered setting `node => node()` in `logger:set_primary_config/2` at boot time. However, RabbitMQ's JSON formatter appends unmapped metadata fields by default unless excluded in `field_map`. Doing it globally would change the JSON log payload for existing users without them opting in. Handling it inside `rabbit_logger_text_fmt` keeps it strictly opt-in for plaintext formats.

Finally, we would document `$node` alongside `$time`, `$level`, `$pid`, and `$msg` in `deps/rabbit/priv/schema/rabbit.schema` and add test cases to `logging_SUITE`.

If this looks like the right direction to you, I can prepare the PR.

**Comment by lukebakken** [comment]
> [!NOTE]
> This comment was written by Claude (Anthropic's Claude Code) under the direction of @lukebakken, who reviewed it before posting. The code references and behaviour claims were verified against `main` at 77f24b0a36 by compiling and running the affected code. The scoping question about JSON and journald below is a genuine open question, not a settled decision.

Thanks for digging into this. The direction is right: a dedicated `format_var/3` clause in `rabbit_logger_text_fmt` is the correct place, and `$node` already parses today, so the formatter clause really is the only missing piece. I verified that through the real config pipeline: `log.console.formatter.plaintext.format = $node $time [$level] $pid $msg` translates to `prefix_format => [node," ",time," [",level,"] ",pid," "]`, and the formatter then renders `<unknown node>` exactly as you describe.

Four corrections before you write the PR.

**Do not read `node` out of `Meta`.** With `maps:get(node, Meta, node())`, any log event whose metadata happens to carry a `node` key prints that value instead of the node that emitted the line, which is the opposite of what this issue asks for. The precedence case you have in mind does not exist: I scanned the abstract code of all 1628 beams in `deps/` for logger metadata, and the complete set of metadata keys in use is `amqp_container`, `channel_number`, `connection`, `domain`, `error_logger`, `exchange`, `pid`, `queue`, `user` and `vhost`. Ra does not attach `node`, and neither does OTP's logger, which adds `gl`, `pid` and `time` plus `mfa`, `line` and `file` from the macros (checked on both 27 and 28). Please use `node()` unconditionally.

**`rabbit_data_coercion:to_list/1` is not needed.** The quoting problem you identified is real, but it comes from the `~tp` in the catch-all clause at `rabbit_logger_text_fmt.erl:56`; `~ts` prints an atom unquoted on its own, including a long Kubernetes name. So the whole clause is:

```erlang
format_var(node, _LogEvent, _Config) ->
    io_lib:format("~ts", [node()]);
```

placed before the catch-all. That also keeps `rabbit_logger_text_fmt` free of a `rabbit_common` call it does not otherwise need.

**There is no `${node?[$node] :}` syntax to fix.** `prepare_fmt_format/2` only ever emits atoms and strings, so the `{Var, Then, Else}` clause at `rabbit_logger_text_fmt.erl:32` is reachable only by setting `prefix_format` directly, which is what `logging_SUITE:setting_message_format_works/1` does at `logging_SUITE.erl:587-598`. Putting `${node?[$node] :}` in a `.conf` file fails config translation today with `{configuration_translation_failure, {badmatch, nomatch}}`. Special-casing `node` in that clause would also make the conditional always take the `Then` branch, so please leave `format_prefix/4` alone.

**Early boot is not a concern here.** The configured formatter is installed by `rabbit_prelaunch_logging:setup/1` from `rabbit.erl:403`, in prelaunch phase 2, which runs after `rabbit_prelaunch_dist:setup/1` at `rabbit_prelaunch.erl:107`. Before that point the default formatter's prefix has no `node` variable at all, so `nonode@nohost` is not a realistic output.

On docs and tests:

- `rabbit.schema` has nowhere to put this. The four `plaintext.format` mappings (lines 1707, 1746, 1801 and 2039) carry only `{default, ...}` and `{datatype, string}`, and the variable list lives at https://www.rabbitmq.com/docs/logging, which currently names exactly `$time`, `$level`, `$pid` and `$msg`. So this needs a companion PR against rabbitmq/rabbitmq-website
- the existing `logging_SUITE` format cases build the format list programmatically through `determine_prefix/1`, and `prepare_fmt_format/1` is not exported, so a case added there would not cover parsing the literal string `$node`. Please add one that goes through `translate_formatter_conf/2`, and assert against `atom_to_list(node())` rather than a fixed name
- a `config_schema_SUITE` snippet would be welcome too. None of the existing snippets set `plaintext.format`

One question I would rather settle before the PR: should the JSON formatter get this as well? Your reason for not adding `node` to primary metadata is correct, and I confirmed it: with the default `field_map`, an unmapped `node` key does land in the JSON output. But `formatter = json` users have the same interleaving problem this issue is about, and journald has no format string at all, so `log.journald.fields = NODE=node` resolves from metadata and comes out empty. Plaintext-only is a reasonable first step if you would rather scope it that way; I just want that to be a decision rather than an omission.

Finally, two pre-existing bugs I ran into while checking this. Please do not fold them into your PR, they should be separate:

- the variable regex at `rabbit_prelaunch_early_logging.erl:325` is `[a-zA_Z0-9_]`, a typo for `[a-zA-Z0-9_]`, so `$NODE` and `$MSG` fail config translation while `$AZ` happens to work
- a `{Var, Then, Else}` element that ends up after `msg` crashes `format_line/5` with `function_clause`, since that function has no tuple clause

**Comment by Vishal-770** [comment]
Hi @lukebakken Thanks for your suggestions, I guess I overcomplicated things somewhat. I am not sure how badly the change will break automatic scrapers reading these logs (json format logs). If it doesn't cause much issue, I am open to adding the same feature to it. Once I raise a PR and when it gets merged, I will create a PR in the RabbitMQ website to update the manual also.

**Comment by Vishal-770** [comment]
Yeah, and for the other two separate issues you mentioned,if  you aren't already working on them, i would like to find some more info about them  and raise an issue for them separately.
